lib/rust/lexer/definition/src/lexer.rs

```diff
@@ -90,89 +104,221 @@ impl EnsoLexer {
+        let token           = comment.into();
+        self.append_token(token);
+        self.pop_states_including(disable_comment);
+        match line_end {
+            token::LineEnding::None => self.on_eof(reader),
+            _                       => {
+                let block_seen_newline = self.block_newline;
+                self.block_state.push_line_ending(line_end);
+                self.block_submit_line(reader);
+                self.push_state(block_seen_newline);
+            },
+        }
     }
 
-    /// Match ASCII alphanumeric characters.
-    fn ascii_alpha_num() -> Pattern {
-        EnsoLexer::ascii_digit() | EnsoLexer::ascii_letter()
+
+    // === Doc Comments ===
+
+    /// Triggered when starting a doc comment.
+    fn on_begin_doc_comment<R:ReaderOps>(&mut self, _reader:&mut R) {
+        trace!(self.logger,"Comment::on_begin_doc_comment");
+        let offset      = self.offset.consume();
+        let indent      = self.consume_current().chars().count();
+        let doc_comment = self.doc_comment;
+        self.comment_state.set_indent(indent);
+        self.comment_state.set_offset(offset);
+        self.push_state(doc_comment);
     }
 
-    /// Match at least one ASCII space character.
-    fn spaces() -> Pattern {
-        Pattern::char(' ').many1()
+    /// Accumulate the contents of the current doc comment line.
+    fn on_build_doc_comment_line<R:ReaderOps>(&mut self, _reader:&mut R) {
+        trace!(self.logger,"Comment::on_build_doc_comment_line");
+        let current = self.consume_current();
+        self.comment_state.append_to_line(current);
     }
 
-    /// Match a newline.
-    ///
-    /// This matches both Unix (LF) and Windows (CRLF) styles of newlines. This is particularly
-    /// important so as not to result in incorrect spans on windows clients.
-    fn newline() -> Pattern {
-        Pattern::char('\n') | Pattern::all_of("\r\n")
+    /// Triggered when a line is ended in a doc comment.
+    fn on_doc_comment_end_of_line<R:ReaderOps>
+    ( &mut self
+    , line_ending:token::LineEnding
+    , reader:&mut R
+    ) {
+        trace!(self.logger,"Comment::on_doc_comment_end_of_line");
+        self.comment_state.submit_line(line_ending);
+        match line_ending {
+            token::LineEnding::None => {
+                let matched_bookmark = self.bookmarks.matched_bookmark;
+                self.bookmarks.rewind(matched_bookmark,reader);
+                self.on_end_doc_comment(reader);
+                self.on_eof(reader);
+            },
+            _ => {
+                let comment_newline = self.doc_comment_newline;
+                self.push_state(comment_newline);
+            }
+        }
     }
 
-    /// The allowable group characters in Enso.
-    fn group_chars() -> String {
-        String::from("()[]{}")
+    /// Triggered when ending a doc comment.
+    fn on_end_doc_comment<R:ReaderOps>(&mut self, _reader:&mut R) {
```

Comment:
I was asking about that in previous PR - why do you pass `_reader` if its not used? Please remove it from all such functions.